// sptc/dist/httpServer.js
const {
	fpm,
	safe_path,
	fileExists,
	getExtension,
	listdir,
	assert,
}=require('../utils')
const {executeSptcFile}=require('../engine')

process.on('uncaughtException', e=>{
  console.log('uncaughtException:', e)
})

function FastCGI_FPM(n, port, locally, option={}) {
	option.env={
		RUNTIME: 'FastCGI_FPM',
	}
	option.workers=n
	fpm(n, _=>FastCGI(port, locally, option))
}

function FastCGI(port, locally, option) {
	const http=require('http')
	const path=require('path')
	const {
		serverDir='.',
		routerEntry=null,
		..._option
	}=option
	_option.env=_option.env || {
		RUNTIME: 'FastCGI',
	}
	_option.srvDir=path.resolve(serverDir)
	_option.router=routerEntry? safe_path(_option.srvDir, routerEntry): null
	http.createServer((req, res)=>{
		CGI(req, res, _option)
	}).listen(port, locally? '127.0.0.1': '0.0.0.0')
}

function CGI(req, res, option) {
	const {
		srvDir,
		exts=['.s', '.sptc', '.sjs'],
		debug=false,
		router=null,
		traverse=false,
		env,
	}=option
	const [reqCtx, handler]=buildRequestContext(req, res, {
		srvDir,
		debug,
		env,
		isRouterMode: router,
	})
	const file=reqCtx.$_REQUEST_FILE || {}
	const fn=router || file.fullname

  function _executeFile(fn) {
	  const ext=getExtension(fn)
  	if(!exts.includes(ext)) {
	  	res.writeHead(403, {'content-type': 'text/plain'})
  		res.end(`extensions does not match: \`${ext}\``)
		  return
	  }
  	executeSptcFile(fn, reqCtx, handler)
  }

	if(fileExists(fn)) {
		_executeFile(fn)
		return
	}

  if(!traverse) {
  	res.writeHead(404, {'content-type': 'text/html'})
	  res.end(`<h3>File does not exist: \`${router || file.pathname}\`</h3>`)
	  return
  }

	for(let e of exts) {
	  if(!fileExists(fn+'/index'+e)) continue
    _executeFile(fn+'/index'+e)
	  return
	}

  res.writeHead(404, {'content-type': 'text/html'})
	res.end(traverseDirectory(srvDir, file.pathname, exts))

}

function buildRequestContext(req, res, {srvDir, debug, env, isRouterMode}) {
	const {fullname, pathname}=getRequestFile(req, srvDir)
	const reqCtx={
		$_RAW_REQUEST: req,
		$_RAW_RESPONSE: res,
		$_ENV: env,
		$_REQUEST_FILE: {
			fullname,
			pathname,
		},
		$_WORKDIR: srvDir,
	}
	reqCtx.$_GLOBAL=reqCtx

	const state={
		statusCode: 200,
		statusText: 'OK',
		responseHeaders: {'content-type': 'text/html; charset=utf8'},
		error: null,
		headerFlushed: false,

		is_debug: debug,
	}
	function flushHeader() {
		if(state.headerFlushed) return;
		state.headerFlushed=true
		res.writeHead(state.statusCode, state.statusText, state.responseHeaders)
	}

	reqCtx.setStatus=(code, text)=>{
		state.statusCode=200
		if(text!==undefined) state.statusText=text
	}
	reqCtx.setResponseHeaders=headers=>{
		Object.assign(state.responseHeaders, headers)
	}
	reqCtx.getResponseStatus=_=>{
		return [state.statusCode, state.statusText]
	}
	reqCtx.getResponseHeaders=_=>{
		return state.responseHeaders
	}

	reqCtx.flushHeader=_=>{
		flushHeader()
	}
	reqCtx.setDebug=x=>{
		state.is_debug=x
	}
	reqCtx.isDebug=_=>state.is_debug
	reqCtx.withCache=createWithCacheStore()

	const handler={
		write: x=>{
			flushHeader()
			res.write(x)
		},
		end: _=>{
			flushHeader()
			res.end()
		},
		onError: e=>{
			if(state.is_debug) console.log(e)
			flushHeader()
			res.end(state.is_debug? e.stack: '')
		},
		isEntry: true,
	}

	return [reqCtx, handler]
}
function getRequestFile(req, srvDir) {
	const _url=require('url')
	const {pathname}=_url.parse(req.url)
	return {
		fullname: safe_path(srvDir, pathname),
		pathname,
	}
}
const cacheStore={}
function createWithCacheStore() {
	const withCache=(fn, k, v)=>{
		assert(k)
		assert(v)
		assert(typeof fn==='function')
	  const o=cacheStore
		if(!o.hasOwnProperty(k) || o[k].v!==v) {
			o[k]={v, x: fn()}
		}
		return o[k].x
	}
	return withCache
}


function traverseDirectory(ref, dir, exts) {
  let ret=`
<!doctype html>
<meta charset=utf8 />
<style>*{line-height: 1.8;}b{font-size: 16px;}a{font-size: 18px; color: #2222ff; cursor: pointer;}a.dir{color: #555555;}</style>`
  const {dirs, files, error}=listdir(ref, dir)
  ret+=`<a style="font-size: 22px;" href="${dir!=='/'? '../': ''}">back</a>`
  ret+=`<h3>${dir}</h3>`
  if(error) {
  	ret+=`<h3>${error.message}</h3>`
  }else{
    for(let i=0; i<dirs.length; i++) {
		  ret+=`<a class="dir" href="${encodeURIComponent(dirs[i].fn)}/">${dirs[i].fn}</a> <br/>`
  	}
  	for(let i=0; i<files.length; i++) {
  		if(exts.includes(getExtension(files[i].fn))) {
  			ret+=`<a href="${encodeURIComponent(files[i].fn)}">${files[i].fn}</a> <br/>`
  		}else{
  	  	ret+=`<b>${files[i].fn}</b> ${files[i].size} <br/>`
  		}
	  }
  }
  return ret
}

module.exports={
	FastCGI,
	FastCGI_FPM,
}
