// sptc/engine/interpreter-macro.js
const {compileSptcMacroFile, MACRO_TYPES}=require('./compiler-macro')

const {
  O_IFDEF,
  O_STR,
  O_DEFINE_CONST,
  O_DEFINE_CALL,
  O_CALL_CONST,
  O_CALL_DEFINE,
  O_DEF,
  O_INCLUDE,
  O_UNDEF,
}=MACRO_TYPES

function buildMacroContext({filename, ...inherits}) {
  const ctx=Object.assign({
    defs: new Set,
    defines: {},
    filename: filename || '',
  }, inherits || {})
  if(filename) {
    ctx.filename=filename
  }
  if(Array.isArray(ctx.defs)) {
    ctx.defs=new Set(ctx.defs)
  }
  return ctx
}

function execute(ctx, ast) {
  let ret=''
  for(let i=0; i<ast.length; i++) {
    const p=ast[i]
    if(p.type===O_IFDEF) {
      const {match, consequent, alternate}=p
      const sub_tree=ctx.defs.has(match)? consequent: alternate
      ret+=execute(ctx, sub_tree)
    }else if(p.type===O_STR) {
      ret+=p.str
    }else if(p.type===O_DEFINE_CONST) {
      ctx.defines[p.cname]=p.cvalue
    }else if(p.type===O_DEFINE_CALL) {
      ctx.defines[p.fname]=p.fcall
    }else if([O_CALL_CONST, O_CALL_DEFINE].includes(p.type)) {
      if(!(p.call in ctx.defines)) {
        ret+=p.source
      }else{
        if(p.type===O_CALL_CONST) {
          ret+=ctx.defines[p.call]
        }else{
          ret+=ctx.defines[p.call](...p.argv)
        }
      }
    }else if(p.type===O_DEF) {
      ctx.defs.add(p.def)
    }else if(p.type===O_UNDEF) {
      ctx.defs.delete(p.def)
    }else if(p.type===O_INCLUDE) {
      const filename=path.resolve(ctx.filename+'/..', p.include)
      const nctx=Object.assign({}, ctx, {filename})
      ret+=executeContext(nctx)
    }else{
      throw new Error('unsupported node: '+JSON.stringify(p))
    }
  }
  return ret
}

function executeMacroContext(ctx, ast) {
  const {filename}=ctx
  return execute(ctx, ast || compileSptcMacroFile(filename))
}

module.exports={
  buildMacroContext,
  executeMacroContext,
}
